r2/service.go

```diff
@@ -59,86 +64,480 @@ func NewService(iOpts instrument.Options, rootPrefix string, store Store) server
+	vars := mux.Vars(r)
+	namespaceID, exists := vars["namespaceID"]
+	if !exists {
+		writeAPIResponse(w, http.StatusBadRequest, "Could not parse namespaceID")
+		return
+	}
+
+	mappingRuleID, exists := vars["ruleID"]
+	if !exists {
+		writeAPIResponse(w, http.StatusBadRequest, "Could not parse ruleID")
+		return
+	}
+
+	hist, err := s.store.FetchMappingRuleHistory(namespaceID, mappingRuleID)
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+	views := make([]mappingRuleJSON, len(hist))
+	for i, mappingRule := range hist {
+		x := mappingRule
+		views[i] = newMappingRuleJSON(x)
+	}
+
+	j, err := json.Marshal(views)
+	if err != nil {
+		writeAPIResponse(w, http.StatusInternalServerError, "Could not create response object")
+		return
+	}
+
+	newPayload(j, http.StatusOK).write(w)
 }
 
-func (s *service) readRollupRule(w http.ResponseWriter, _ *http.Request) {
-	w.WriteHeader(http.StatusNotImplemented)
-	fmt.Fprintf(w, "Placeholder for readRollupRule")
+func (s *service) readRollupRule(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	namespaceID, exists := vars["namespaceID"]
+	if !exists {
+		writeAPIResponse(w, http.StatusBadRequest, "Could not parse namespaceID")
+		return
+	}
+
+	rollupRuleID, exists := vars["ruleID"]
+	if !exists {
+		writeAPIResponse(w, http.StatusBadRequest, "Could not parse rollupRuleID")
+		return
+	}
+
+	rollupRule, err := s.store.FetchRollupRule(namespaceID, rollupRuleID)
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+	view := newRollupRuleJSON(rollupRule)
+
+	j, err := json.Marshal(view)
+	if err != nil {
+		writeAPIResponse(w, http.StatusInternalServerError, "Could not create response object")
+		return
+	}
+
+	newPayload(j, http.StatusOK).write(w)
 }
 
-func (s *service) createRollupRule(w http.ResponseWriter, _ *http.Request) {
-	w.WriteHeader(http.StatusNotImplemented)
-	fmt.Fprintf(w, "Placeholder for createRollupRule")
+func (s *service) createRollupRule(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	namespaceID, exists := vars["namespaceID"]
+	if !exists {
+		writeAPIResponse(w, http.StatusBadRequest, "Could not parse namespaceID")
+		return
+	}
+
+	var rrj rollupRuleJSON
+	errs := parseRequest(&rrj, r.Body)
+	if len(errs) > 0 {
+		writeAPIResponse(w, http.StatusBadRequest, errs)
+		return
+	}
+
+	rollupRule, err := s.store.CreateRollupRule(namespaceID, rrj.rollupRuleView())
+
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+
+	view := newRollupRuleJSON(rollupRule)
+
+	j, err := json.Marshal(view)
+	if err != nil {
+		writeAPIResponse(w, http.StatusInternalServerError, "Could not create response object")
+		return
+	}
+
+	newPayload(j, http.StatusCreated).write(w)
 }
 
-func (s *service) updateRollupRule(w http.ResponseWriter, _ *http.Request) {
-	w.WriteHeader(http.StatusNotImplemented)
-	fmt.Fprintf(w, "Placeholder for updateRollupRule")
+func (s *service) updateRollupRule(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
```

Comment:
you seem to have a lot of similar arg checking in place, might be useful to write a utility helper - 

e.g. 

```
func parseVars(r *http.Request, names ...string) (map[string]string, error) {
   vars := mux.Vars(r)
   for _, n := range names {
     if _, ok := vars[n]; !ok {
       return nil, fmt.Errorf("Could not parse %s", n)
     }
   }
   return vars, nil
}

func (s *service) updateRollupRule(w http.ResponseWriter, r *http.Request) {
    vars, err := parseVars(r, "namespaceID", "ruleID")
    if err != nil {
       writeAPIResponse(w, http.StatusBadRequest, err.Error())
    }
    // ... 
} 
```

Reply:
I don't actually need this validation. I'm only checking if its set to something. If it isn't set to anything the regex for the route matching wouldn't work. Can just get rid of it.